### ISSUE TITLE
Enhance healthcheck to detect runtime issues with package dependencies

### ISSUE THREAD

**Issue by szelenka** [PrefectHQ/prefect#2902]
## Description
Dependency issue with urllib3 not caught during healthcheck.

The Flow can run successfully on a local machine, and even build/deploy to Prefect Cloud.

The problem is that when Prefect Cloud attempts to run this Flow, the Flow Run will reach an error during the scheduling. In our case, a K8 Agent picks up that the Flow should run, and spawns a Pod to run the Flow. However, immediately after the Flow Run starts, it'll dump these errors to the Pod console (the Flow Run is stuck in a starting state, because the Pod cannot write logs back to Prefect Cloud):
```
[2020-07-01 13:40:32] DEBUG - prefect.CloudFlowRunner | Failed to retrieve flow state with error: AttributeError("'SSLSocket' object has no attribute 'connection'")
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Failed to write log with error: 'SSLSocket' object has no attribute 'connection'
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Failed to write log with error: 'SSLSocket' object has no attribute 'connection'
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Failed to write log with error: 'SSLSocket' object has no attribute 'connection'
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Failed to write log with error: 'SSLSocket' object has no attribute 'connection'
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Unable to write logs to Prefect Cloud
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Unable to write logs to Prefect Cloud
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Unable to write logs to Prefect Cloud
[2020-07-01 13:40:32] CRITICAL - CloudHandler | Unable to write logs to Prefect Cloud
```

For this specific failure, we had `snowflake-connector-python==2.2.8` specified, which seems to have upgrades the `urlli3` version to `1.25.9` and that appears to have an API breaking change introduced. Rolling back to `snowflake-connector-python==2.2.7` mitigates this specific problem.

But this version dependency was not caught during the healthcheck.

## Expected Behavior
When deploying the Flow, I would expect the healthcheck to detect any problems with the code/dependencies before deploying it. In this specific example, the Flow would run and build on a LocalEnvironment because it never had to reach out to PrefectCloud to log anything. But when deploying it through PrefectCloud, it would attempt to write the logs back but reach an exception (which was due to a dependency problem).

## Reproduction
```python
from prefect import Flow, task
from prefect.utilities.logging import get_logger
from prefect.environments import LocalEnvironment
from prefect.environments.storage import Docker
from prefect.engine.executors import LocalDaskExecutor
from prefect.engine.results import S3Result


@task
def sample():
    get_logger().info('Executing Task')


with Flow(
    name="Prefect Bug",
    storage=Docker(
        registry_url='containers.local/test_org',
        base_image='containers.local/test_org/prefect:0.12.1-python3.8',
        python_dependencies=[
            'ujson==3.0.0',
            'requests==2.23.0',
            'pandas==1.0.5',
            'numpy==1.19.0',
            'sqlalchemy==1.3.18',
            'hvac==0.10.4',
            'snowflake-connector-python==2.2.8',
            'snowflake-sqlalchemy==1.2.3',
            'pyarrow==0.17.1',
        ],
    ),
    environment=LocalEnvironment(
        executor=LocalDaskExecutor(
            scheduler='threads',
            num_workers=8
        ),
        labels=["test"]
    ),
    result=S3Result(
        bucket='prefect-flow-results',
        boto3_kwargs=dict(
            region_name='us-east-1',
            endpoint_url='https://minio.local/',
        )
    )
) as flow:
    result = sample()

if __name__ == "__main__":
    flow.register(
        project_name="test",
        build=True
    )

```

## Environment
```json
{
  "config_overrides": {
    "cloud": {
      "use_local_secrets": true
    },
    "context": {
      "secrets": false
    }
  },
  "env_vars": [],
  "system_information": {
    "platform": "macOS-10.15.5-x86_64-i386-64bit",
    "prefect_version": "0.12.1",
    "python_version": "3.8.1"
  }
}
```

**Comment by joshmeek** [comment]
Thanks for opening @szelenka! I'm wondering how we could ensure that the dependencies function properly for communication with the set API backend without making a direct query because this would require the API access to be configured during the build step. 🤔

**Comment by szelenka** [comment]
Since it needs to reach out to the server to register the Flow, could we perform the register from within the Docker image it just built? 

It wouldn't catch the problem during the healthcheck, but it would expose the problem at registration/build time.

**Comment by cicdw** [comment]
Two points to add:

This feels like a slippery slope - for example, this doesn't feel far off from detecting:
```python
@task
def my_broken_task():
    print(unscoped_undefined_variable)
```
which becomes increasingly complicated.

Second (more positive) point: if we allow for `.py` file-based storage within `Docker` storage (#2887 ) we could run the script within the container which would detect the broken import in this particular example without introducing complexity.

**Comment by joshmeek** [comment]
@cicdw Going to enhance the healthcheck script in #2944 to account for this 👍

**Comment by szelenka** [comment]
I ran into another issue, similar to this ... where I wanted to do a basic Flow storing the results to S3Result. 

Turns out, I need to manually specify that I want `botocore` in order for it to use the S3Reusult class from Prefect. In most cases I have a base docker image with that stuff baked in already, but this time I was not using a base image. The image built, deployed, and ran fine .. until it needed to use S3Result when it said, "hey I don't have botocore, tough luck!"

Would the enhanced healthcheck would catch that type of dependency error as well?
